File: mobile/packages/lighchat_firebase/lib/src/e2ee/heal_session.dart
/// Lazy self-heal E2EE sessions на мобайле.
///
/// Зеркало `src/lib/e2ee/v2/heal-session.ts`, назначение — закрыть те же
/// регрессии:
///  1. Новое устройство появилось в `users/{uid}/e2eeDevices`, но текущая
///     session-doc (до ротации) не содержит для него wrap — ничего не
///     расшифровывается.
///  2. Session-doc с неподдерживаемым `protocolVersion` (legacy v1 или
///     неизвестной — мы такие просто переротируем в v2).
///  3. Участник чата не виден в session.wraps (например, добавили новое
///     устройство после создания эпохи).
///
/// Heal идемпотентный: параллельные вызовы в пределах одного процесса
/// дедуплицируются in-memory, а Firestore transaction защищает от двойной
/// ротации между разными клиентами.
///
/// Модуль не тянет Riverpod / UI слой — только `cloud_firestore` и наши
/// существующие e2ee-примитивы. Его можно вызывать из `MobileE2eeRuntime`
/// (mobile/app), из фоновых задач или из тестов.
library;

import 'package:cloud_firestore/cloud_firestore.dart';

import 'device_identity.dart';
import 'session_firestore.dart';
import 'system_events.dart';
import 'telemetry.dart';

/// Почему потребовался heal. Возвращается в `HealResult`, логируется в
/// телеметрии — удобно дебажить.
enum MobileHealReason {
  noSessionDoc,
  sessionUnsupported,
  myDeviceMissing,
  otherDeviceMissing,
}

class MobileHealResult {
  const MobileHealResult._({
    required this.healed,
    required this.newEpoch,
    this.reason,
  });

  factory MobileHealResult.noop() =>
      const MobileHealResult._(healed: false, newEpoch: 0);

  factory MobileHealResult.rotated({
    required MobileHealReason reason,
    required int newEpoch,
  }) =>
      MobileHealResult._(healed: true, newEpoch: newEpoch, reason: reason);

  final bool healed;
  final int newEpoch;
  final MobileHealReason? reason;
}

/// Диагностика без побочных эффектов. Если нет current epoch (ещё не включали
/// E2EE) — возвращает `noop`: это не heal-кейс, а enable-кейс.
Future<MobileHealResult> diagnoseMobileSessionCoverage({
  required FirebaseFirestore firestore,
  required String conversationId,
  required int currentEpoch,
  required List<String> participantIds,
  required String currentUserId,
  required MobileDeviceIdentityV2 identity,
}) async {
  if (currentEpoch < 1) return MobileHealResult.noop();

  final fetched = await fetchE2eeSessionAny(
    firestore: firestore,
    conversationId: conversationId,
    epoch: currentEpoch,
  );
  if (fetched == null) {
    return MobileHealResult.rotated(
      reason: MobileHealReason.noSessionDoc,
      newEpoch: currentEpoch + 1,
    );
  }
  if (!fetched.isV2) {
    // Session неподдерживаемого формата (legacy v1 / unknown). Ротируем в v2.
    return MobileHealResult.rotated(
      reason: MobileHealReason.sessionUnsupported,
      newEpoch: currentEpoch + 1,
    );
  }
  final session = fetched.v2!;
  final perUser = session.wraps[currentUserId] ?? const {};
  if (!perUser.containsKey(identity.deviceId)) {
    return MobileHealResult.rotated(
      reason: MobileHealReason.myDeviceMissing,
      newEpoch: currentEpoch + 1,
    );
  }
  try {
    final bundles = await collectParticipantDevices(
      firestore: firestore,
      participantIds: participantIds,
    );
    for (final entry in bundles.entries) {
      final wrapped = (session.wraps[entry.key] ?? const {}).keys.toSet();
      for (final dev in entry.value) {
        if (!wrapped.contains(dev.deviceId)) {
          return MobileHealResult.rotated(
            reason: MobileHealReason.otherDeviceMissing,
            newEpoch: currentEpoch + 1,
          );
        }
      }
    }
  } on E2eeSessionException {
    // Нет устройства у кого-то из участников — не heal-кейс.
    return MobileHealResult.noop();
  }
  return MobileHealResult.noop();
}

/// In-memory guard для конкурентных вызовов. Ключ — `$cid:$epoch`.
final Map<String, Future<MobileHealResult>> _inflightHeal =
    <String, Future<MobileHealResult>>{};

/// Главная точка входа: убеждается, что current epoch покрывает все активные
/// устройства участников. Если нет — ротирует эпоху.
///
/// Best-effort: ошибки не пробрасываются наружу, возвращается `noop`. Это
/// отличие от `createE2eeSessionDocV2`, которая должна кидать, но heal —
/// optimistic self-heal, а не операция, которую UI должен принудительно
/// завершить.
Future<MobileHealResult> healSessionForCurrentDevices({
  required FirebaseFirestore firestore,
  required String conversationId,
  required int currentEpoch,
  required List<String> participantIds,
  required String currentUserId,
  required MobileDeviceIdentityV2 identity,
}) async {
  final cacheKey = '$conversationId:$currentEpoch';
  final existing = _inflightHeal[cacheKey];
  if (existing != null) return existing;

  final future = () async {
    try {
      // Публикуем текущее устройство до ротации — если это новый девайс /
      // пост-восстановление, без этого новая эпоха снова не покроет нас.
      await ensureMobileDevicePublished(
        firestore: firestore,
        userId: currentUserId,
        identity: identity,
      );

      final diag = await diagnoseMobileSessionCoverage(
        firestore: firestore,
        conversationId: conversationId,
        currentEpoch: currentEpoch,
        participantIds: participantIds,
        currentUserId: currentUserId,
        identity: identity,
      );
      if (!diag.healed) return MobileHealResult.noop();

      // Транзакция: атомарно читаем текущий epoch и обновляем его. Если
      // другой клиент успел раньше — откатимся к `noop` (пусть вызывающий
      // перечитает актуальный epoch из conversation).
      final convRef =
          firestore.collection('conversations').doc(conversationId);
      int nextEpoch = currentEpoch + 1;
      var conflict = false;

      await firestore.runTransaction((tx) async {
        final snap = await tx.get(convRef);
        if (!snap.exists) {
          throw const E2eeSessionException('E2EE_CONV_NOT_FOUND');
        }
        final raw = snap.data()?['e2eeKeyEpoch'];
        final latest =
            raw is int ? raw : (raw is num ? raw.toInt() : 0);
        if (latest > currentEpoch) {
          conflict = true;
          return;
        }
        nextEpoch = latest + 1;
        tx.update(convRef, <String, Object?>{
          'e2eeKeyEpoch': nextEpoch,
          'e2eeEnabled': true,
          'e2eeEnabledAt': DateTime.now().toUtc().toIso8601String(),
        });
      });

      if (conflict) return MobileHealResult.noop();

      final bundles = await collectParticipantDevices(
        firestore: firestore,
        participantIds: participantIds,
      );
      await createE2eeSessionDocV2(
        firestore: firestore,
        conversationId: conversationId,
        epoch: nextEpoch,
        currentIdentity: identity,
        currentUserId: currentUserId,
        participantDevices: bundles,
      );

      logE2eeEvent(
        E2eeTelemetryEventType.rotateSuccess,
        E2eeTelemetryPayload(
          userId: currentUserId,
          conversationId: conversationId,
          deviceId: identity.deviceId,
          metrics: {'epoch': nextEpoch},
        ),
      );
      try {
        await ChatSystemEventFactories.epochRotated(
          firestore: firestore,
          conversationId: conversationId,
          epoch: nextEpoch,
          actorUserId: currentUserId,
        );
      } catch (_) {
        // Timeline-маркер — не критично.
      }

      return MobileHealResult.rotated(
        reason: diag.reason ?? MobileHealReason.myDeviceMissing,
        newEpoch: nextEpoch,
      );
    } catch (e) {
      logE2eeEvent(
        E2eeTelemetryEventType.rotateFailure,
        E2eeTelemetryPayload(
          userId: currentUserId,
          conversationId: conversationId,
          deviceId: identity.deviceId,
          errorCode: e.toString(),
        ),
      );
      return MobileHealResult.noop();
    } finally {
      _inflightHeal.remove(cacheKey);
    }
  }();
  _inflightHeal[cacheKey] = future;
  return future;
}

File: src/lib/e2ee/v2/heal-session.ts
'use client';

/**
 * Lazy session re-keying для multi-device self-healing.
 *
 * Зачем: исходный дизайн предусматривал ротацию эпохи только на явных событиях
 * (enable, add/remove member, revoke, manual). В реальном мире пользователь
 * входит на новое устройство или чистит кэш, и новая `e2eeDevices/{id}` запись
 * появляется **без** триггера ротации. В результате:
 *  1. Сообщения от других участников приходят только на те устройства, что
 *     были зарегистрированы на момент предыдущей ротации — новый девайс не в
 *     `wraps[me]`.
 *  2. После очистки кэша браузер получает новый `deviceId`/publicKey и теряет
 *     возможность unwrap'ить chat-key старой эпохи.
 *  3. На send получение chat-key падает → `E2EE_NO_CHAT_KEY`, UI показывает
 *     «Не удалось зашифровать сообщение».
 *
 * Fix: клиент **сам** определяет рассогласование и триггерит новую эпоху.
 *
 * Рассогласование обнаруживается так:
 *  - session-doc для текущего `e2eeKeyEpoch` не существует / v1 (legacy), а
 *    текущий клиент не v1 → нужно создать v2 эпоху.
 *  - В v2 session `wraps[me][identity.deviceId]` отсутствует → мой девайс не
 *    покрыт → rotate.
 *  - У любого участника активных devices больше, чем ключей в `wraps[uid]` →
 *    чужой девайс не покрыт → rotate (чтобы остальные участники могли читать
 *    с новых устройств).
 *
 * После ротации:
 *  - `conversations/{id}.e2eeKeyEpoch` инкрементируется,
 *  - новый session-doc содержит wraps для всех текущих активных устройств.
 *
 * Все операции идемпотентные: если другой клиент уже ротирнул, мы просто
 * перечитываем документ и работаем с новой эпохой.
 *
 * Изоляция: модуль не зависит от send/receive pipeline напрямую — его можно
 * вызывать из любого места (на mount чата, перед send, после ошибки unwrap).
 */

import {
  doc,
  getDoc,
  runTransaction,
  type Firestore,
} from 'firebase/firestore';
import type { Conversation, E2eeSessionDocV2 } from '@/lib/types';
import {
  collectParticipantDevicesV2,
  createE2eeSessionDocV2,
  fetchE2eeSessionAny,
} from '@/lib/e2ee/v2/session-firestore-v2';
import {
  getOrCreateDeviceIdentityV2,
  publishE2eeDeviceV2,
  type DeviceIdentityV2,
} from '@/lib/e2ee/v2/device-identity-v2';
import {
  chatSystemEvents,
  postChatSystemEventV2,
} from '@/lib/e2ee/v2/system-events';
import { logE2eeEvent, normalizeErrorCode } from '@/lib/e2ee/v2/telemetry';
import { logger } from '@/lib/logger';

export type HealReason =
  | 'no-session-doc'
  | 'session-unsupported'
  | 'my-device-missing'
  | 'other-device-missing';

export type HealResult =
  | { healed: true; reason: HealReason; newEpoch: number }
  | { healed: false };

/**
 * Определяет, нужна ли ротация. Чтение-онли, без побочных эффектов.
 */
export async function diagnoseSessionCoverageV2(
  firestore: Firestore,
  conversation: Conversation,
  currentUserId: string,
  currentIdentity: DeviceIdentityV2
): Promise<{ needsHeal: boolean; reason: HealReason | null; currentEpoch: number }> {
  const currentEpoch = conversation.e2eeKeyEpoch ?? 0;
  if (currentEpoch < 1) {
    // Эпохи ещё нет — это отдельный сценарий (enable), здесь не heal'им.
    return { needsHeal: false, reason: null, currentEpoch };
  }

  const session = await fetchE2eeSessionAny(firestore, conversation.id, currentEpoch);
  if (!session) {
    return { needsHeal: true, reason: 'no-session-doc', currentEpoch };
  }
  if (session.version !== 'v2') {
    // Неопознанный / legacy session-doc. Перекатываем эпоху в v2.
    return { needsHeal: true, reason: 'session-unsupported', currentEpoch };
  }
  const data: E2eeSessionDocV2 = session.data;
  const myWraps = data.wraps?.[currentUserId];
  if (!myWraps || !myWraps[currentIdentity.deviceId]) {
    return { needsHeal: true, reason: 'my-device-missing', currentEpoch };
  }

  // Участники: сравним set активных devices с тем, что покрыто в wraps.
  // Если у кого-то появились новые devices — перекодируем, чтобы они смогли
  // читать сообщения.
  try {
    const participants = conversation.participantIds ?? [];
    const bundles = await collectParticipantDevicesV2(firestore, participants);
    for (const b of bundles) {
      const wrappedIds = new Set(Object.keys(data.wraps?.[b.userId] ?? {}));
      for (const dev of b.devices) {
        if (!wrappedIds.has(dev.deviceId)) {
          return { needsHeal: true, reason: 'other-device-missing', currentEpoch };
        }
      }
    }
  } catch {
    // collectParticipantDevicesV2 может бросить E2EE_NO_DEVICE(uid). Это не
    // heal-кейс — это ошибка данных, эскалируем выше в send-pipeline.
    return { needsHeal: false, reason: null, currentEpoch };
  }

  return { needsHeal: false, reason: null, currentEpoch };
}

/**
 * In-memory гвард, чтобы два параллельных вызова healSession в пределах одной
 * вкладки не сгенерировали две конкурирующие ротации. Транзакция на backend
 * защитит от дубликата на уровне Firestore, но UX улучшится, если мы просто
 * подождём первую попытку.
 */
const inflightHeal = new Map<string, Promise<HealResult>>();

/**
 * Пытается привести session в консистентное состояние. Идемпотентно:
 * повторные вызовы для той же эпохи либо возвращают `{ healed: false }`
 * (если уже покрыто), либо результат предыдущего in-flight вызова.
 */
export async function healSessionForCurrentDevicesV2(
  firestore: Firestore,
  conversation: Conversation,
  currentUserId: string
): Promise<HealResult> {
  const cacheKey = `${conversation.id}:${conversation.e2eeKeyEpoch ?? 0}`;
  const existing = inflightHeal.get(cacheKey);
  if (existing) return existing;

  const promise = (async (): Promise<HealResult> => {
    try {
      const identity = await getOrCreateDeviceIdentityV2();
      // На всякий случай сначала опубликуем наш публичник — если это новое
      // устройство / cache-clear, он ещё не в `e2eeDevices`, и без этого
      // новая эпоха не покроет нас.
      await publishE2eeDeviceV2(firestore, currentUserId, identity);

      const diag = await diagnoseSessionCoverageV2(
        firestore,
        conversation,
        currentUserId,
        identity
      );
      if (!diag.needsHeal) return { healed: false };

      // Всё: ротируем. Используем transaction для чтения e2eeKeyEpoch и для
      // атомарного обновления — если другой клиент успел раньше, мы
      // прочитаем его значение и откажемся от своей ротации.
      const convRef = doc(firestore, 'conversations', conversation.id);
      let nextEpoch = diag.currentEpoch + 1;
      let conflict = false;

      await runTransaction(firestore, async (tx) => {
        const snap = await tx.get(convRef);
        if (!snap.exists()) throw new Error('E2EE_CONV_NOT_FOUND');
        const latestEpoch = (snap.data()?.e2eeKeyEpoch as number | undefined) ?? 0;
        if (latestEpoch > diag.currentEpoch) {
          // Кто-то уже ротирнул между нашим diagnose и txn start.
          conflict = true;
          return;
        }
        nextEpoch = latestEpoch + 1;
        tx.update(convRef, {
          e2eeKeyEpoch: nextEpoch,
          e2eeEnabled: true,
          e2eeEnabledAt: new Date().toISOString(),
        });
      });

      if (conflict) {
        // Не делаем ничего — другой клиент создал session. Следующий read
        // возьмёт её. Возвращаем healed=false, чтобы вызывающий не думал,
        // что это мы ротировали.
        return { healed: false };
      }

      // Теперь создаём session-doc для нашей новой эпохи. Transaction выше
      // зафиксировала, что nextEpoch уникален для нас — safe to setDoc.
      const participants = conversation.participantIds ?? [];
      const bundles = await collectParticipantDevicesV2(firestore, participants);
      await createE2eeSessionDocV2(
        firestore,
        conversation.id,
        nextEpoch,
        identity,
        currentUserId,
        bundles
      );

      logE2eeEvent('e2ee.v2.rotate.success', {
        userId: currentUserId,
        conversationId: conversation.id,
        deviceId: identity.deviceId,
        metrics: { epoch: nextEpoch },
      });

      try {
        await postChatSystemEventV2({
          firestore,
          conversationId: conversation.id,
          event: chatSystemEvents.epochRotated(nextEpoch, currentUserId),
        });
      } catch (e) {
        logger.warn('e2ee-heal', 'system-event post failed', e);
      }

      return { healed: true, reason: diag.reason ?? 'my-device-missing', newEpoch: nextEpoch };
    } catch (e) {
      logE2eeEvent('e2ee.v2.rotate.failure', {
        userId: currentUserId,
        conversationId: conversation.id,
        errorCode: normalizeErrorCode(e),
      });
      // Не пробрасываем наверх — heal best-effort. Вызывающий получит
      // `{ healed: false }` и покажет стандартный error-placeholder.
      logger.warn('e2ee-heal', 'heal failed', e);
      return { healed: false };
    } finally {
      inflightHeal.delete(cacheKey);
    }
  })();
  inflightHeal.set(cacheKey, promise);
  return promise;
}

/**
 * Отдельная кнопка / CTA в UI: «починить шифрование». Делает то же, что
 * `healSessionForCurrentDevicesV2`, но всегда пытается ротировать (даже если
 * diagnose говорит, что всё ок). Полезно, когда пользователь видит ошибку
 * расшифровки, но автоматическая эвристика не сработала.
 */
export async function forceRotateEpochV2(
  firestore: Firestore,
  conversation: Conversation,
  currentUserId: string
): Promise<{ newEpoch: number }> {
  const identity = await getOrCreateDeviceIdentityV2();
  await publishE2eeDeviceV2(firestore, currentUserId, identity);

  const convRef = doc(firestore, 'conversations', conversation.id);
  const snap = await getDoc(convRef);
  if (!snap.exists()) throw new Error('E2EE_CONV_NOT_FOUND');
  const latestEpoch = (snap.data()?.e2eeKeyEpoch as number | undefined) ?? 0;
  const nextEpoch = latestEpoch + 1;

  const participants = conversation.participantIds ?? [];
  const bundles = await collectParticipantDevicesV2(firestore, participants);
  await createE2eeSessionDocV2(
    firestore,
    conversation.id,
    nextEpoch,
    identity,
    currentUserId,
    bundles
  );
  await runTransaction(firestore, async (tx) => {
    const s = await tx.get(convRef);
    if (!s.exists()) throw new Error('E2EE_CONV_NOT_FOUND');
    const cur = (s.data()?.e2eeKeyEpoch as number | undefined) ?? 0;
    if (cur >= nextEpoch) return; // кто-то раньше — OK, наш doc валиден
    tx.update(convRef, {
      e2eeKeyEpoch: nextEpoch,
      e2eeEnabled: true,
      e2eeEnabledAt: new Date().toISOString(),
    });
  });

  logE2eeEvent('e2ee.v2.rotate.success', {
    userId: currentUserId,
    conversationId: conversation.id,
    deviceId: identity.deviceId,
    metrics: { epoch: nextEpoch },
  });
  try {
    await postChatSystemEventV2({
      firestore,
      conversationId: conversation.id,
      event: chatSystemEvents.epochRotated(nextEpoch, currentUserId),
    });
  } catch (e) {
    logger.warn('e2ee-heal', 'force-rotate: system-event post failed', e);
  }
  return { newEpoch: nextEpoch };
}
